File: examples/rust/examples/client_chat_basic.rs
//! Basic Chat Example
//!
//! This example demonstrates basic chat functionality with ekoDB's AI integration.
//! It shows how to send a simple chat message with context search.
//!
//! Prerequisites:
//! - Set OPENAI_API_KEY, ANTHROPIC_API_KEY, or PERPLEXITY_API_KEY in .env
//! - Run the ekoDB server: `make run`
//!
//! Run with: `cargo run --example client_chat_basic`

use ekodb_client::{
    ChatMessageRequest, Client, CollectionConfig, CreateChatSessionRequest, Record,
};
use std::error::Error;

#[tokio::main]
async fn main() -> Result<(), Box<dyn Error>> {
    // Load environment variables
    dotenv::dotenv().ok();

    println!("=== ekoDB Chat Basic Example ===\n");

    // Create client
    let client = Client::builder()
        .base_url(
            std::env::var("API_BASE_URL").unwrap_or_else(|_| "http://localhost:8080".to_string()),
        )
        .api_key(std::env::var("API_BASE_KEY")?)
        .build()?;

    let collection = "client_chat_basic_rust";

    // Step 1: Insert some sample data for context
    println!("=== Inserting Sample Data ===");
    let mut doc1 = Record::new();
    doc1.insert("title", "Introduction to ekoDB");
    doc1.insert("content", "ekoDB is a high-performance database with intelligent caching, real-time capabilities, and AI integration.");
    doc1.insert("category", "documentation");
    client.insert(collection, doc1).await?;

    let mut doc2 = Record::new();
    doc2.insert("title", "Search Features");
    doc2.insert("content", "ekoDB supports full-text search, vector search, and hybrid search with automatic context retrieval.");
    doc2.insert("category", "features");
    client.insert(collection, doc2).await?;

    let mut doc3 = Record::new();
    doc3.insert("title", "AI Chat Integration");
    doc3.insert("content", "The chat feature allows you to query your database using natural language and get AI-powered responses with relevant context.");
    doc3.insert("category", "features");
    client.insert(collection, doc3).await?;

    println!("✓ Inserted 3 sample documents\n");

    // Step 2: Create a chat session
    println!("=== Creating Chat Session ===");
    let session_request = CreateChatSessionRequest {
        collections: vec![CollectionConfig {
            collection_name: collection.to_string(),
            fields: vec![],
            search_options: None,
        }],
        llm_provider: "openai".to_string(),
        llm_model: Some("gpt-4".to_string()),
        system_prompt: Some("You are a helpful assistant for ekoDB documentation.".to_string()),
        parent_id: None,
        branch_point_idx: None,
        max_context_messages: Some(10),
        bypass_ripple: Some(false),
    };

    let session = client.create_chat_session(session_request).await?;
    let chat_id = &session.chat_id;
    println!("✓ Created session: {}\n", chat_id);

    // Step 3: Send a chat message
    println!("=== Sending Chat Message ===");
    let message_request = ChatMessageRequest::new("What is ekoDB and what features does it have?");
    let response = client.chat_message(chat_id, message_request).await?;

    println!("Message ID: {}", response.message_id);
    println!("\n=== AI Response ===");
    for (i, resp) in response.responses.iter().enumerate() {
        println!("Response {}: {}", i + 1, resp);
    }

    println!(
        "\n=== Context Used ({} snippets) ===",
        response.context_snippets.len()
    );
    for (i, snippet) in response.context_snippets.iter().enumerate() {
        println!("\nSnippet {}:", i + 1);
        println!("  Collection: {}", snippet.collection);
        println!("  Score: {:.4}", snippet.score);
        println!("  Matched Fields: {:?}", snippet.matched_fields);
        println!("  Record: {:?}", snippet.record);
    }

    println!("\nExecution Time: {}ms", response.execution_time_ms);

    // Cleanup: Delete the collection (chat session is managed by server)
    println!("\n=== Cleanup ===");
    client.delete_collection(collection).await?;
    println!("✓ Deleted collection");

    println!("\n✓ Chat completed successfully");

    Ok(())
}

File: examples/kotlin/examples/ClientChatBasic.kt
package io.ekodb.client.examples

import io.ekodb.client.EkoDBClient
import io.ekodb.client.types.Record
import kotlinx.coroutines.runBlocking
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.put
import kotlinx.serialization.json.putJsonArray
import kotlinx.serialization.json.putJsonObject

/**
 * Basic Chat example - Simple AI chat with context search
 */
fun main() = runBlocking {
    val baseUrl = System.getenv("API_BASE_URL") ?: "http://localhost:8080"
    val apiKey = System.getenv("API_BASE_KEY") ?: error("API_BASE_KEY environment variable not set")
    
    val client = EkoDBClient.builder()
        .baseUrl(baseUrl)
        .apiKey(apiKey)
        .build()
    
    val collection = "kotlin_chat_basic_example"
    
    println("=== ekoDB Kotlin Client - Basic Chat Example ===\n")
    
    try {
        // Step 1: Insert sample data for context
        println("=== Inserting Sample Data ===")
        
        val doc1 = Record.new()
            .insert("title", "Introduction to ekoDB")
            .insert("content", "ekoDB is a high-performance database with intelligent caching, real-time capabilities, and AI integration.")
            .insert("category", "documentation")
        client.insert(collection, doc1)
        
        val doc2 = Record.new()
            .insert("title", "Search Features")
            .insert("content", "ekoDB supports full-text search, vector search, and hybrid search with automatic context retrieval.")
            .insert("category", "features")
        client.insert(collection, doc2)
        
        val doc3 = Record.new()
            .insert("title", "AI Chat Integration")
            .insert("content", "The chat feature allows you to query your database using natural language and get AI-powered responses with relevant context.")
            .insert("category", "features")
        client.insert(collection, doc3)
        
        println("✓ Inserted 3 sample documents\n")
        
        // Step 2: Create a chat session
        println("=== Creating Chat Session ===")
        
        val sessionRequest = buildJsonObject {
            putJsonArray("collections") {
                add(buildJsonObject {
                    put("collection_name", collection)
                    putJsonArray("fields") {}
                })
            }
            put("llm_provider", "openai")
            put("llm_model", "gpt-4")
            put("system_prompt", "You are a helpful assistant for ekoDB documentation.")
            put("max_context_messages", 10)
            put("bypass_ripple", false)
        }
        
        val session = client.createChatSession(sessionRequest)
        val chatId = session["chat_id"]?.jsonPrimitive?.content ?: error("No chat_id in response")
        println("✓ Created session: $chatId\n")
        
        // Step 3: Send a chat message
        println("=== Sending Chat Message ===")
        
        val messageRequest = buildJsonObject {
            put("message", "What is ekoDB and what features does it have?")
        }
        
        val response = client.chatMessage(chatId, messageRequest)
        println("✓ Chat response:")
        println("  Message ID: ${response["message_id"]}")
        println("  Responses: ${response["responses"]}\n")
        
        // Step 4: Cleanup
        println("=== Cleanup ===")
        client.deleteChatSession(chatId)
        println("✓ Deleted chat session")
        
    } finally {
        try {
            client.deleteCollection(collection)
            println("✓ Deleted collection: $collection")
        } catch (e: Exception) {
            println("⚠ Could not delete collection: ${e.message}")
        }
        
        client.close()
        println("\n✓ Basic chat example completed successfully")
    }
}
